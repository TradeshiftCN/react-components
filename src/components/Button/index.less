@import '../../style/variables/index';

@keyframes ts-rc-move-progress-bar {
	0% {
		background-position: 0 0;
	}
	100% {
		background-position: 40px 40px;
	}
}

.ts-rc-button {
	display: inline-block;
	padding-left: @ts-unit;
	padding-right: @ts-unit;
	border: 1px solid transparent;
	line-height: @ts-unit-double - 2px;
	border-radius: @ts-unit;
	transition: @ts-transition-now;
	font-size: @ts-fontsize;
	position: relative;
	&:not(:disabled):hover {
		text-decoration: none;
	}
	& + & {
		margin-left: 5px;
	}
	&--small {
		padding-left: @ts-unit-half;
		padding-right: @ts-unit-half;
		line-height: @ts-unit - 2px;
		font-size: @ts-fontsize-small;
	}
	&--default {
		border-color: @ts-color-gray-light;
		color: @ts-color-black;
		&:hover {
			background-color: @ts-color-gray-lighter;
		}
	}
	&--primary {
		background-color: @ts-color-blue;
		border-color: @ts-color-blue;
		color: @ts-color-white;
		&:hover {
			background-color: @ts-color-blue-dark;
		}
	}
	&--warning {
		background-color: @ts-color-orange;
		color: @ts-color-white;
		border-color: @ts-color-white;
	}
	&--danger {
		background-color: @ts-color-red;
		color: @ts-color-white;
		border-color: @ts-color-white;
	}
	&--circle {
		padding-left: 0;
		padding-right: 0;
		width: @ts-unit-double;
		border-radius: @ts-unit;
	}
	&--loading {
		&,
		&:hover {
			background-color: transparent;
			border-color: transparent;
			cursor: default;
		}
		& > * {
			visibility: hidden;
		}
		&::after {
			content: '';
			display: block;
			position: absolute;
			visibility: visible;
			left: 0;
			top: @ts-unit-half;
			right: 0;
			bottom: @ts-unit-half;
			background-image: linear-gradient(
				-45deg,
				hsla(198, 20%, 83%, 0.65) 25%,
				hsla(200, 21%, 92%, 0.65) 25%,
				hsla(200, 21%, 92%, 0.65) 50%,
				hsla(198, 20%, 83%, 0.65) 50%,
				hsla(198, 20%, 83%, 0.65) 75%,
				hsla(200, 21%, 92%, 0.65) 75%,
				hsla(200, 21%, 92%, 0.65)
			);
			background-size: @ts-unit-double @ts-unit-double;
			border-radius: inherit;
			animation: ts-rc-move-progress-bar 1s linear infinite;
		}
	}
	& > span {
		text-transform: uppercase;
	}
	& > i {
		font-size: @ts-fontsize-icon;
		vertical-align: middle;
	}
	& > * + * {
		margin-left: 8px;
	}
}
