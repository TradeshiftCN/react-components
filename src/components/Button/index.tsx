import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import _ from 'lodash';

export type ButtonProps = {
	type?: 'default' | 'primary' | 'danger' | 'warning';
	className?: string;
	loading?: boolean;
	size?: 'small';
	children?: React.ReactNode;
	icon?: string;
	label?: string;
	disabled?: boolean;
} & React.DetailedHTMLProps<React.HTMLAttributes<HTMLButtonElement>, HTMLButtonElement>;

const prefixCls = 'ts-rc-button';

function genButtonLabel(children: React.ReactNode) {
	return React.Children.map(children, child => {
		if (_.isString(child)) {
			return <span>{child}</span>;
		}

		return child;
	});
}

function Button(props: ButtonProps) {
	const { type, className, loading, size, children, icon, label, disabled, ...buttonProps } = props;
	const labelNodes = React.Children.count(children) ? children : label;
	const hasOnlyIcon = !_.isNil(icon) && _.isNil(labelNodes);
	if (_.isNil(labelNodes) && _.isNil(icon)) {
		return null;
	}

	if (size === 'small' && icon) {
		throw new Error("small size can't used with icon");
	}

	return (
		<button
			{...buttonProps}
			className={cx(prefixCls, className, `${prefixCls}--${type}`, {
				[`${prefixCls}--${size}`]: !!size,
				[`${prefixCls}--circle`]: hasOnlyIcon,
				[`${prefixCls}--loading`]: !!loading
			})}
			disabled={!!loading || !!props.disabled}
		>
			{!!icon && <i className={`ts-icon-${icon}`} />}
			{!_.isNil(labelNodes) && genButtonLabel(labelNodes)}
		</button>
	);
}

Button.propTypes = {
	type: PropTypes.oneOf(['default', 'primary', 'danger', 'warning']),
	loading: PropTypes.bool,
	className: PropTypes.string,
	size: PropTypes.oneOf(['small']),
	children: PropTypes.node,
	icon: PropTypes.string,
	label: PropTypes.string,
	disabled: PropTypes.bool
};

Button.defaultProps = {
	type: 'default',
	loading: false
};

export default Button;
