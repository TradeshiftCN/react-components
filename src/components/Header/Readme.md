### Basic Usage

```jsx
initialState = { theme: 'white' };

<div>
	<Header
		title="Title"
		icon="logo.svg"
		theme={state.theme}
		backButton={true}
		forwardButton={{
			onClick() {
				alert('click forward button');
			}
		}}
	/>
	<div style={{ marginTop: '16px' }}>
		Change theme:
		{['white', 'blue'].map(v => (
			<label style={{ margin: '0 5px', cursor: 'pointer' }} key={v}>
				{v}{' '}
				<input
					type="radio"
					name="theme"
					onChange={e => setState({ theme: v })}
					value={v}
					checked={v === state.theme}
				/>
			</label>
		))}
	</div>
</div>;
```

### Customize elements

```jsx
<Header
	title="Title"
	icon="logo.svg"
	backButton={<button>&lt;-back</button>}
	forwardButton={<button>forward-&gt;</button>}
/>
```

### Buttons

_Note: buttons aside wasn't implemented_

```jsx
import { Button } from '@tradeshift/react-components';
<Header
	title="Title"
	icon="logo.svg"
	buttons={[<Button type="primary">primary</Button>, <Button>default</Button>]}
/>;
```

### Tabs

_Note: TabPane shouldn't has children when working together with Header_

```jsx
import { Tabs } from '@tradeshift/react-components';
const { TabPane } = Tabs;
initialState = { activeTab: '2' };

<div>
	<Header
		title="Title"
		icon="logo.svg"
		tabs={
			<Tabs onChange={key => setState({ activeTab: key })} activeKey={state.activeTab}>
				<TabPane tab="One" key="1" />
				<TabPane tab="Two" key="2" />
				<TabPane tab="Three" key="3" />
			</Tabs>
		}
	/>
	<p style={{ marginTop: '16px' }}>active tab key: {state.activeTab}</p>
</div>;
```

### Search

```jsx
import { Search } from '@tradeshift/react-components';

<Header
	title="Title"
	icon="logo.svg"
	search={
		<Search onSearch={v => alert(`Your input is '${v}'`)} placeholder="Press enter to search" />
	}
/>;
```
