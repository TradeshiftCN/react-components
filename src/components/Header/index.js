import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import './index.less';

class Header extends React.Component {
  render() {
    const { className, children, icon, title, ...other } = this.props;
    return (
      <header className={cx(`${prefixCls}-header`, className)} {...other}>
        <ul className={`${prefixCls}-header__bars`}>
          <li className={`${prefixCls}-header__bars-tool`}>
            <menu className={`${prefixCls}-header__bars-tool-item`}>
              <li className={`${prefixCls}-header__bars-tool-item-icon`}>
                <img src={icon} alt="icon" width={40} height={40} />
              </li>
              <li className={`${prefixCls}-header__bars-tool-item-title`}>
                <label>{title}</label>
              </li>
            </menu>
          </li>
          {children && <li>{children}</li>}
        </ul>
      </header>
    );
  }
}

Header.propTypes = {
  className: PropTypes.string,

  children: PropTypes.node,

  // Header 左侧的 icon
  icon: PropTypes.string,

  // Header 标题
  title: PropTypes.string
};

export default Header;
