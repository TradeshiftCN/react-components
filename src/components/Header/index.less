@import '~ui-variables';
@import '~ui-mixins';

@width: 100%;
@height: @ui-unit-triple;

.@{prefix-cls}-header {
  position: relative;
  z-index: @ui-zindex-topbar;
  width: 100%;
  background-color: @ui-color-white;
  border-bottom: 1px solid @ui-color-gray-light;
  .ui-mixin-transition(opacity);
  &__bars {
    &-tool {
      position: relative;
      user-select: none;
      height: @height - 1px;
      white-space: nowrap;
      &-item {
        position: absolute;
        font-size: 0;
        white-space: nowrap;
        margin: 0 !important;
        > li {
          font-size: @ui-fontsize;
          display: inline-block;
          vertical-align: middle;
          padding: @ui-unit-half @ui-unit-quarter;
        }
        li&-icon {
          .ui-mixin-reduce-l();
          height: @height;
          max-height: @height;
          min-height: @height;
        }
        &-title {
          label {
            font-size: 18px;
            padding: 10px;
            padding-left: 0;
            line-height: 20px;
            display: block;
            position: relative;
            top: -1px;
            overflow: hidden;
            text-overflow: ellipsis;
          }
        }
      }
    }
  }
}
