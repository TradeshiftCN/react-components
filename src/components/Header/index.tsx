import React from 'react';
import cx from 'classnames';
import _ from 'lodash';
import PropTypes from 'prop-types';

type HTMLButtonElementProps = React.HTMLProps<HTMLButtonElement> &
	React.HTMLAttributes<HTMLButtonElement>;
type NavigationBtn = React.ReactNode | HTMLButtonElementProps | boolean;

export type HeaderProps = {
	title?: string | React.ReactNode;
	icon?: string;
	theme?: 'blue' | 'white';
	tabs?: React.ReactNode;
	search?: React.ReactNode;
	backButton?: NavigationBtn;
	forwardButton?: NavigationBtn;
	buttons?: React.ReactNode[];
	className?: string;
};

type NavigationBtnConfig = {
	key: string;
	icon: string;
	btn: NavigationBtn;
	onClick(): void;
};

const prefixCls = 'ts-rc-header';

function genNavigationBtns(
	className: string,
	backButton?: NavigationBtn,
	forwardButton?: NavigationBtn
) {
	const navigationBtns: React.ReactNode[] = _.compact(
		([
			{
				key: 'back',
				btn: backButton,
				icon: 'ts-icon-arrowleft',
				onClick: () => {
					window.history.back();
				}
			},
			{
				key: 'forward',
				btn: forwardButton,
				icon: 'ts-icon-arrowright',
				onClick: () => {
					window.history.forward();
				}
			}
		] as NavigationBtnConfig[]).map(v => {
			if (!v.btn) {
				return;
			}

			if (React.isValidElement(v.btn)) {
				return (
					<li className={className} key={v.key}>
						{v.btn}
					</li>
				);
			}

			const props =
				v.btn === true
					? {
							onClick: v.onClick
					  }
					: v.btn;

			return (
				<li className={className} key={v.key}>
					<button {...props}>
						<i className={v.icon}></i>
					</button>
				</li>
			);
		})
	);

	return navigationBtns;
}

export function Header(props: HeaderProps) {
	const genItemClassName = (className?: string) => {
		return cx(`${prefixCls}-item`, { [prefixCls + '-item-' + className]: !!className });
	};
	const navigationBtns = genNavigationBtns(
		genItemClassName('navigation'),
		props.backButton,
		props.forwardButton
	);

	return (
		<header className={cx(prefixCls, `${prefixCls}--${props.theme}`, props.className)}>
			<menu className={`${prefixCls}-main`}>
				{!!navigationBtns.length && navigationBtns}
				{!!props.icon && (
					<li className={genItemClassName('icon')}>
						<img src={props.icon} alt="icon" className={`${prefixCls}-icon`} />
					</li>
				)}
				{!!props.title && (
					<li className={genItemClassName('title')}>
						{React.isValidElement(props.title) ? props.title : <h1>{props.title}</h1>}
					</li>
				)}
				{Array.isArray(props.buttons) &&
					React.Children.map(
						props.buttons.slice().reverse(),
						btn => !_.isNil(btn) && <li className={genItemClassName('button')}>{btn}</li>
					)}
			</menu>
			{!!props.tabs && <section className={`${prefixCls}-tabs`}>{props.tabs}</section>}
			{!!props.search && <section className={`${prefixCls}-search`}>{props.search}</section>}
		</header>
	);
}

Header.propTypes = {
	title: PropTypes.node,
	icon: PropTypes.string,
	theme: PropTypes.oneOf(['blue', 'white']),
	tabs: PropTypes.node,
	search: PropTypes.node,
	/** if set to `true`, this will call `window.history.back()` */
	backButton: PropTypes.oneOfType([PropTypes.bool, PropTypes.object, PropTypes.node]),
	/** if set to `true`, this will call `window.history.forward()` */
	forwardButton: PropTypes.oneOfType([PropTypes.bool, PropTypes.object, PropTypes.node]),
	buttons: PropTypes.arrayOf(PropTypes.node),
	className: PropTypes.string
};

Header.defaultProps = {
	theme: 'white'
};

export default Header;
