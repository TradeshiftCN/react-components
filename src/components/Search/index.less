@import '../../style/variables/index';

.ts-rc-search {
	position: relative;
	line-height: @ts-unit-double;
	i {
		position: absolute;
		top: 0;
		display: inline-block;
		width: @ts-unit-double;
		height: @ts-unit-double;
		line-height: inherit;
		vertical-align: middle;
		font-size: @ts-fontsize-icon;
	}
	input {
		line-height: @ts-unit-double - 2px !important;
		padding: 0 @ts-unit-double;
		border: 1px solid @ts-color-gray-light;
		width: 100%;
		display: block;
		border-radius: @ts-unit;
		&:hover,
		&:focus {
			box-shadow: 0 0 0 1px @ts-color-blue-light;
		}
		&:hover {
			border-color: @ts-color-blue-lighter;
		}
		&:focus {
			border-color: @ts-color-blue;
		}
	}
	&-search-icon {
		// FIXME: the baseline of this icon has a little offset
		padding-top: 1px;
		left: 0;
		font-weight: @ts-fontweight-semibold;
	}
	&-remove-icon {
		right: 0;
		color: @ts-color-slate-lightest;
		cursor: pointer;
	}
}
