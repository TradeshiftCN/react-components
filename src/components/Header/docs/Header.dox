/**
 * @title 基本功能
 * @renderModel left
 */
import Header from '@tradeshift/react-components/lib/Header'

const HeaderBasic = () => {
  return (
    <Header icon="/svg/appLogo.svg" title="Header" />
  )
}

@component Header
