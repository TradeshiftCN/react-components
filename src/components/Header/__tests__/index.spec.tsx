import React from 'react';
import { shallow } from 'enzyme';
import Header from '..';
import Button from '../../Button';
import Search from '../../Search';
import Tabs from '../../Tabs';

describe('Header', () => {
	it('basic', () => {
		const wrapper = shallow(
			<Header
				title="title"
				tabs={<Tabs />}
				search={<Search />}
				buttons={[<Button>btn 1</Button>, <Button>btn 2</Button>]}
				backButton={true}
				forwardButton={<button>forward</button>}
			/>
		);

		expect(wrapper).toMatchSnapshot();
	});
});
