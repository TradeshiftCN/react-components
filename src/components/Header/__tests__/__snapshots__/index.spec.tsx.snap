// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header basic 1`] = `
<header
  className="ts-rc-header ts-rc-header--white"
>
  <menu
    className="ts-rc-header-main"
  >
    <li
      className="ts-rc-header-item ts-rc-header-item-navigation"
      key="back"
    >
      <button
        onClick={[Function]}
      >
        <i
          className="ts-icon-arrowleft"
        />
      </button>
    </li>
    <li
      className="ts-rc-header-item ts-rc-header-item-navigation"
      key="forward"
    >
      <button>
        forward
      </button>
    </li>
    <li
      className="ts-rc-header-item ts-rc-header-item-title"
    >
      <h1>
        title
      </h1>
    </li>
    <li
      className="ts-rc-header-item ts-rc-header-item-button"
      key=".0"
    >
      <Button
        loading={false}
        type="default"
      >
        btn 2
      </Button>
    </li>
    <li
      className="ts-rc-header-item ts-rc-header-item-button"
      key=".1"
    >
      <Button
        loading={false}
        type="default"
      >
        btn 1
      </Button>
    </li>
  </menu>
  <section
    className="ts-rc-header-tabs"
  >
    <Tabs
      destroyInactiveTabPane={true}
      renderTabBar={[Function]}
      renderTabContent={[Function]}
    />
  </section>
  <section
    className="ts-rc-header-search"
  >
    <Search />
  </section>
</header>
`;
